### ISSUE TITLE
Learning about point tasks in a process

### ISSUE THREAD

**Issue by opensdh** [StanfordLegion/legion#1746]
Using global variables and the usual thread synchronization, it is easy to take process-local notes as point tasks execute.  However, the distribution of point tasks over (Unix) processes is in general known only by the relevant mapper instances, and it could be expensive to communicate it to other processes where the point tasks end up running.  As such, no process knows when all its assigned point tasks have executed (so as to finalize/transmit the collected notes).  (As a particular example, a process might end up with 0 point tasks such that there is no hint as to when the launch took place at all.)

Could we have a means of learning from Legion (which already has to arrange for the right number of calls to take place) either how many such point tasks from a particular (index) launch have or will run in each process or that, again per-launch, no more point tasks will be started in each process?  One form that this interface might take would be a callback function (with access at least to the `TaskArgument`) called once per `Legion::Runtime` per task launch.  It would either provide the number as an argument or (following the CPS idea) be invoked after all relevant point tasks.  In the usual (non-debugging) situation of one `Runtime` per process, this produces one call on each process as desired, independent of sharding patterns.

A toy example might be to collect per-process statistics about a field for lightweight _in-situ_ analysis:
```
std::atomic<unsigned> max;
void bunny_task(...) {
  unsigned mine=0;
  for(i...) {
    // update various fields, including bunny_field
    mine=std::max(mine,bunny_field[i]);
  }
  max.fetch_max(mine);
}
void callback(Legion::Task *) {
  if(max.exchange(0)>75) std::cerr << getpid() << ": too many bunnies\n";
}
void caller_task() {
  run->execute_index_space(ctx,bunny_task_launcher);
}
void near_main() {
  Runtime::set_top_level_task_id(...);
  Runtime::register_launch_complete_callback(callback);
  Runtime::start(...);
}
```
Obviously as written this works only if all the launches of `bunny_task` are serialized (though their point tasks can run in parallel!).  Passing the `Legion::Task*` to the callback allows a more sophisticated (read: realistic) callback to perform separate bookkeeping for separate launches.

A very different CPS-like approach (which would be sufficient for _our_ use cases) might be to equip `Legion::Future` with a `then` (and `Legion::FutureMap` with an `all_then`) that performs a callback (in the process that registered the callback, or else in all processes that collectively registered one) when the future is (all the futures are) ready without requiring a separate task launch or blocking.

**Comment by opensdh** [comment]
One interesting special case here is a task configured with `set_idempotent`: there the _total_ number of point tasks is not known _a priori_, although it would seem that the same interface would suffice.

**Comment by lightsighter** [comment]
> A very different CPS-like approach (which would be sufficient for our use cases) might be to equip Legion::Future with a then (and Legion::FutureMap with an all_then) that performs a callback (in the process that registered the callback, or else in all processes that collectively registered one) when the future is (all the futures are) ready without requiring a separate task launch or blocking.

How is that different than launching a task which only depends on that future? Tasks are already callbacks that you get when dependences are satisfied. If you don't want to send the task all the way through the pipeline you can even make it a `local_function_task` since it has no region requirements, but that is an optimization and not required.
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h?ref_type=heads#L1619-1627

**Comment by opensdh** [comment]
A `local_function_task` might be most of what we want here, but I'm not sure:
1. We would want to be able to have it proceed only when an entire `FutureMap` was complete (ideally more efficiently than by adding every `FutureMap::get_future` result to `TaskLauncher::futures`).
1. Does "on a local processor where the parent task is executing" mean that the local function task runs on _every_ such processor when the parent task is control-replicated, or just on an unspecified one of them?
1. What exactly does "pure function with no side effects" mean here?  Without any region requirements, of course such a task cannot alter any field values, so is this a statement about making other calls into Legion (_e.g._, executing further tasks) or what?

**Comment by lightsighter** [comment]
> We would want to be able to have it proceed only when an entire FutureMap was complete (ideally more efficiently than by adding every FutureMap::get_future result to TaskLauncher::futures).

You can't do this at the moment, but it probably wouldn't be too hard to add. Do you really want it to wait on every future in the future map (all point tasks), or are just certain futures enough?

> Does "on a local processor where the parent task is executing" mean that the local function task runs on every such processor when the parent task is control-replicated, or just on an unspecified one of them?

It will actually run on exactly the same processor as the parent task (when the parent is preempted and the futures have all completed).

> What exactly does "pure function with no side effects" mean here? Without any region requirements, of course such a task cannot alter any field values, so is this a statement about making other calls into Legion (e.g., executing further tasks) or what?

Mostly this means that you don't have any region requirements. You can have side effects, but like calling `printf` but it's up to you to manage them in that case. I think you can have local function tasks launch further sub-tasks, but since they can't have any region arguments so there is no way to pass privileges.

**Comment by opensdh** [comment]
> You can't do this at the moment, but it probably wouldn't be too hard to add. Do you really want it to wait on every future in the future map (all point tasks), or are just certain futures enough?

Well, for the particular use case we have in mind, it would be sufficient to wait on all the futures corresponding to point tasks dispatched by the same `Runtime` or by the same shard.  I said just "all" because it seemed more likely to be generically useful.

> It will actually run on exactly the same processor as the parent task (when the parent is preempted and the futures have all completed).

So, to be sure, if the parent task is control-replicated and (as one would expect) all of its shards post the same local function task, that task runs once each on every processor that has been running (one shard of) the parent task?

> Mostly this means that you don't have any region requirements. You can have side effects, but like calling `printf` but it's up to you to manage them in that case. I think you can have local function tasks launch further sub-tasks, but since they can't have any region arguments so there is no way to pass privileges.

I suppose you could create entirely new logical regions?

**Comment by lightsighter** [comment]
> Well, for the particular use case we have in mind, it would be sufficient to wait on all the futures corresponding to point tasks dispatched by the same Runtime or by the same shard. I said just "all" because it seemed more likely to be generically useful.

Right, that is an optimization, but you would probably want to express it as a dependence on the whole future map.

> So, to be sure, if the parent task is control-replicated and (as one would expect) all of its shards post the same local function task, that task runs once each on every processor that has been running (one shard of) the parent task?

Yes, we implicitly replicate local function tasks so we run a copy on every shard if the parent is control replicated. This is easy to do since the only inputs and outputs are futures and not regions.

> I suppose you could create entirely new logical regions?

You could I suppose, but I think we might enforce that the selected variant be a leaf task variant, so you could make the new logical regions but then not populate it with any data.

**Comment by opensdh** [comment]
So it sounds like `TaskLauncher::local_function_task` with a new `TaskLauncher::future_maps` or so would satisfy our requirements, so long as `set_idempotent` doesn't spoil it.  Can we have that?

**Comment by lightsighter** [comment]
> So it sounds like TaskLauncher::local_function_task with a new TaskLauncher::future_maps or so would satisfy our requirements, so long as set_idempotent doesn't spoil it. Can we have that?

What do you mean by "so long as `set_idempotent` doesn't spoil it?" What are you afraid that setting it idempotent would do?

**Comment by opensdh** [comment]
I can imagine that idempotent task launches, since the runtime has permission to run them extra times, might complicate the accounting: might a `Future`(`Map`) returned from an idempotent task launch become ready (in the `wait_all_results` sense as appropriate) before _all_ point tasks have completed if the value is known from those that have?

**Comment by lightsighter** [comment]
> might a Future(Map) returned from an idempotent task launch become ready (in the wait_all_results sense as appropriate) before all point tasks have completed if the value is known from those that have?

No, `wait_all_results` can only return once all the futures are ready. If one point task has to restart for resilience reasons then it won't be considered complete and therefore not be counted towards `wait_all_results`. Similarly a `Future` for a particular point task is not considered complete until that point is done running. If the point fails and is restarted for resilience reasons (why you might mark it idempotent) then the future will only make it to the complete stage once we know we no longer need to rollback and rerun the task.

**Comment by opensdh** [comment]
I was under the impression that idempotence might be used not for resilience but just to substitute computation for communication, but for the purpose of this issue that doesn't matter: if the `FutureMap` becomes ready only when all the point tasks (idempotent extras or no) are complete, then the `TaskLauncher::future_maps` idea is good enough.  Is that something we can have soon?

**Comment by lightsighter** [comment]
Just thinking about what we've got already, can you use `Runtime::reduce_future_map` to reduce the future map down to a a future that you then pass as a precondition to the `TaskLauncher`? I'm not just suggesting this for convenience; I also suspect that it will be better for performance as well because it will do a butterfly all-reduce between the shards for communicating when all the futures are done which is more scalable than what you'll get with the `FutureMap::wait_all_results` call today.

**Comment by opensdh** [comment]
We would need to use `reduce_future_map` on the user tasks' `FutureMap` objects for which we don't necessarily have any reductions defined (most commonly because the task returns `void`).  Is there some dummy `ReductionOpID` that could be used to not actually reduce anything?  Would the original values (when not `void`) still be available via the `FutureMap`?

**Comment by lightsighter** [comment]
> You can't have function parameters of type void, nor can you ever have references to void, so I don't see how to do this.

I've always found that a bit ridiculous. Rust's type system has the nice bottom type/value `()`.

> Does FutureMap::get_void_result work even when the actual type isn't void?

Yes, it works regardless of the return type.

> Could we avoid this circumlocution by having Legion provide a LEGION_REDOP_VOID that does nothing, or would that be just as hard (though in Legion)?

Let me think about this. In general, Legion's philosophy is that dependences should only exist if they are data dependences. This would go against that philosophy because we'd only be propagating a control dependence and not any actual data.

> Also, the question about the destructiveness or not of the reduction remains.

I've lost track of this: what is the question about destructiveness?

**Comment by opensdh** [comment]
> I've always found that a bit ridiculous. Rust's type system has the nice bottom type/value `()`.

It has of course [been proposed](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0146r1.html) to rectify that, but obviously compatibility is often an insurmountable challenge.  (But it's not the usual [bottom](https://en.wikipedia.org/wiki/Bottom_type), but [unit](https://en.wikipedia.org/wiki/Unit_type).)

> > Could we avoid this circumlocution by having Legion provide a LEGION_REDOP_VOID that does nothing, or would that be just as hard (though in Legion)?
> 
> Let me think about this. In general, Legion's philosophy is that dependences should only exist if they are data dependences. This would go against that philosophy because we'd only be propagating a control dependence and not any actual data.

I'm not sure that's a coherent position.  Trivially, we already have `get_void_result`, and obviously that's supposed to establish a synchronization without a data dependency.  At a slightly more complicated level, it suggests that as soon as one `bool` is known to be `true` in a `LEGION_REDOP_OR_BOOL` reduction, we can launch tasks that depend on the result, denying an obvious place a user might expect synchronization.  (We could say we do want that, but then what about `INT_MAX` with `LEGION_REDOP_MAX_INT32`?  Is that special enough that it should defeat synchronization?)

Moreover, "true" data dependencies are [very hard to define](https://open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3064r2.pdf), and appealing to the fact that Legion (not being a compiler) can't conduct any of the proofs required to make that definition interesting is just saying that `void` (construed as a unit type) is special just because we happen to know its value statically.  Would we then extend that special treatment to `std::nullptr_t`?  To `std::monostate`?  To any type satisfying `std::is_empty_v`?  To `std::tuple<std::nullptr_t,std::nullptr_t>`?

> > Also, the question about the destructiveness or not of the reduction remains.
> 
> I've lost track of this: what is the question about destructiveness?

From a few posts above:

> Would the original values (when not void) still be available via the FutureMap?

**Comment by lightsighter** [comment]
> I'm not sure that's a coherent position. Trivially, we already have get_void_result, and obviously that's supposed to establish a synchronization without a data dependency.

That's a synchronization and implies blocking behavior. It's not an asynchronous dependence. Legion supports the blocking behavior because it has to, but there are no ways to express asynchronous control dependences in Legion.

> At a slightly more complicated level, it suggests that as soon as one bool is known to be true in a LEGION_REDOP_OR_BOOL reduction, we can launch tasks that depend on the result, denying an obvious place a user might expect synchronization. (We could say we do want that, but then what about INT_MAX with LEGION_REDOP_MAX_INT32? Is that special enough that it should defeat synchronization?)

We don't do short-circuiting with reduction operators because it leads to races with the distributed reduction protocols. We always run the full protocol even if technically we could short-circuit a computation. This also ensures that reduction operators can be treated like black boxes and Legion doesn't need to know the semantics of them. Legion doesn't explicitly know about the default reduction operators which are really only there as a library of convenience for users.

> Moreover, "true" data dependencies are [very hard to define](https://open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3064r2.pdf),

It may be hard to define for multi-thread code with a relaxed memory model, but not for "single-threaded" programs with sequential semantics. As with other (apparently) sequential programs, all Legion programs can trivially define true data dependences: every read has to see the result of the most recent write (and the accumulation of any reductions on top of that most recent write) as defined by the program order of the tasks. Memory coherence and memory consistency models are both trivial in such circumstances and true data dependences can be [precisely defined (unless you use relaxed coherence modes in which case it is not trivial but still well defined)](https://legion.stanford.edu/pdfs/oopsla2013.pdf).

> Would we then extend that special treatment to std::nullptr_t? To std::monostate? To any type satisfying std::is_empty_v? To std::tuple<std::nullptr_t,std::nullptr_t>?

No, those are different than `void` because they are types with non-empty storage and information can be passed through those types. You should have no trouble defining reduction operators using those types and Legion will abide by them. As long as there are bits where information can flow then Legion has to respect that as a true data dependence.

> Would the original values (when not void) still be available via the FutureMap?

Yes, future maps are always immutable and are not consumed by an all-reduce operation.

**Comment by opensdh** [comment]
> That's a synchronization and implies blocking behavior. It's not an asynchronous dependence. Legion supports the blocking behavior because it has to, but there are no ways to express asynchronous control dependences in Legion.

I'm not sure I know what "asynchronous control dependence" means.  Are you saying that a `void` `Legion::Future` can't _delay_ the launch of a task via `TaskLauncher.futures` (which I think I remember discussing at some point)?  If so, sure, but then we just get into `std::nullptr_t` and so forth again and we wonder why `void` is special.

> It may be hard to define for multi-thread code with a relaxed memory model, but not for "single-threaded" programs with sequential semantics.

I'm not sure that's the right way to read that discussion.  The _effects_ of data dependencies on the analysis of relaxed-memory-model code are certainly subtle, but the definition doesn't ever seem easy to me.  Does `y` depend on `x` in `y=a*x+b`?  What if `a` is 0&mdash;for a particular evaluation, or throughout a particular task, or via a template parameter, or via a macro, or as a literal?  What if `a` isn't 0, but is a sufficiently large power of 2 that the multiplication with `x` is shifted out of the smaller range of `y` and is lost in the conversion?  What if `b` was previously assigned `c-a*x` such that it's just `y=c`?

> No, those are different than `void` because they are types with non-empty storage and information can be passed through those types. You should have no trouble defining reduction operators using those types and Legion will abide by them. As long as there are bits where information can flow then Legion has to respect that as a true data dependence.

I don't know what sort of information you mean if it can be passed through `std::nullptr_t`: "reading" such an object doesn't even touch memory&nbsp;([conv.lval]/3.1).  It's true that its `sizeof` is non-zero, so you can talk about its object representation as a byte buffer, but it sounds strange to use a C++ type not for its values but for its memory, but since `sizeof(void)` doesn't exist at all we might as well pretend that it's 1 (with all padding bits) like compilers do for their `void*` arithmetic extensions as say that it uniquely is 0.

Or are you saying that Legion is interested in the object representation as a byte buffer?  I suppose that works even with the reduction interface, since everything is done by reference, but it seems more than a little fragile to have `x=y` break your data because it doesn't actually copy anything.

> > Would the original values (when not void) still be available via the FutureMap?
> 
> Yes, future maps are always immutable and are not consumed by an all-reduce operation.

Great.  I think that means the question about `void` is the only thing open now.  (If Legion decides not to support it, I suppose my second choice would be to use an internal empty enumeration type, so that we could simply map it back to `void` for our users; I _think_ that's easy to implement.)

**Comment by lightsighter** [comment]
> Are you saying that a void Legion::Future can't delay the launch of a task via TaskLauncher.futures (which I think I remember discussing at some point)?

Yes.

> If so, sure, but then we just get into std::nullptr_t and so forth again and we wonder why void is special.

Legion doesn't understand types. Futures that carry no data (have zero-sized buffers, aka `void` futures) are special because no information (in the information theoretic sense) can be passed by value through them (although I suppose you could abuse them by delaying setting them and communicate information in a side-channel way through how long it takes for your task to finish and set the future, but that is an egregious abuse of the programming model).

> Does y depend on x in y=a*x+b? What if a is 0—for a particular evaluation, or throughout a particular task, or via a template parameter, or via a macro, or as a literal? What if a isn't 0, but is a sufficiently large power of 2 that the multiplication with x is shifted out of the smaller range of y and is lost in the conversion? What if b was previously assigned c-a*x such that it's just y=c?

Legion doesn't care. If there are bits that are passed from one computation (aka task) to another then there is a data dependence; no bits means no data dependence. The interpretation of those bits is up to the application. Legion won't optimize away for things that C++ might consider undefined behavior (overflow, etc). Moving bits -> data dependence.

> Or are you saying that Legion is interested in the object representation as a byte buffer?

Precisely.

>  I suppose that works even with the reduction interface, since everything is done by reference, but it seems more than a little fragile to have x=y break your data because it doesn't actually copy anything.

Not sure what you mean by this.

**Comment by opensdh** [comment]
> > Are you saying that a void Legion::Future can't delay the launch of a task via TaskLauncher.futures (which I think I remember discussing at some point)?
> 
> Yes.

I guess a significant part of what (little) remains to discuss is why Legion bothers to notice this case: if what Legion cares about is byte buffers, then `void` is special only in that the buffer is empty (rather than being special in the type system), and I can't immediately think of a reason that a client would express a dependency graph in terms of `void` futures and be happy that Legion ignored those dependencies.

> Legion doesn't care. If there are bits that are passed from one computation (aka task) to another then there is a data dependence; no bits means no data dependence. The interpretation of those bits is up to the application. Legion won't optimize away for things that C++ might consider undefined behavior (overflow, etc). Moving bits -> data dependence.

I guess I would call this a "potential data dependence", since it's really a statement about the limited knowledge Legion has about the application.

> > I suppose that works even with the reduction interface, since everything is done by reference, but it seems more than a little fragile to have x=y break your data because it doesn't actually copy anything.
> 
> Not sure what you mean by this.

In
```cpp
std::nullptr_t x,y;
std::memset(&x,1,1);
y=x;
assert(std::memchr(&y,1,1)==&y);
```
the assert can fail because the byte stored in the "buffer" `x` is not copied to `y` by the assignment.  That's the cost of summarily ignoring the type system, which might or might not be acceptable.

**Comment by opensdh** [comment]
We now have the mapper-based approach working with tracing (by using `select_task_options` and ignoring our own results when running a trace other than the first time), but I realize that that brings back the potential for trouble from idempotence.  How does restarting (or, if it happens, duplication) of point tasks work there?  Surely our wrapper code would run again, which is bad unless we can learn that it's happening.

**Comment by lightsighter** [comment]
Documenting our discussion from the meeting today: currently marking a task as `idempotent` does not have any impact on whether a task can be re-run or not in Legion. It is an aspirational feature only at the moment. In the future, it will be used to know that a task could potentially be able to be rerun either for resilience purposes or because of mis-speculation during speculative execution. Both of those features will be opt-in features though so applications and mappers will know when they are being used, and the runtime won't just be suddenly rerunning tasks out of thin air. This means situations involving idempotent tasks being rerun are safe to ignore for now and will only become relevant if, and only if, you actually start using future resilience and/or speculative execution features.
